### PR TITLE
Add new user manually

### PR DESCRIPTION
**Pull request by denuwanthi**

### CODE REVIEW

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.hbs [hunk 1]
```diff
@@ -47,31 +78,11 @@
...
-                </div>
-            </div>
-
-
-            <input class="btn btn-primary add-margin-bottom-3x" type="submit" name="addUser" value="Add User"/>
+            {{#defineZone "accountVerificationMethod-area"}}
+                <!-- Here is generated content for different account verification ui's when adding a user -->
+            {{/defineZone}}
+            <input class="btn btn-primary add-margin-bottom-3x" type="submit" name="addUser" id="addUser"
+                   value="Add User"/>
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.hbs]: Support Internationalization

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js [hunk 2]
```diff
@@ -2,17 +2,67 @@ function onGet(env) {
...
+}
 function getDomainNames(env) {
     var domainNames;
     if (env.config.isDomainInLogin) {
         try {
             domainNames = callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
                 "getDomainNames", []);
         } catch (e) {
-            return {errorMessage: 'signup.error.retrieve.domain'};
+            sendError(500, "user.add.error.retrieve.domain");
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: Use other error code than http error codes

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/PasswordPolicyBean.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.is.portal.user.client.api.bean;
+
+/**
+ * This class contains the password policy bean.
+ * TODO: Add annotations once migrated to C5
+ */
+public class PasswordPolicyBean {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/bean/PasswordPolicyBean.java]: add toString() method to bean

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    });
+    $('#verificationSelector').change(displayVals);
+});
+$(window).load(function () {
+    jQuery.validator.addMethod("noSpace", function (value, element) {
+        return value.indexOf(" ") < 0 ;
+    }, "Please don't enter spaces");
+});
+function displayVals() {
+    if ($("#verificationSelector option:selected").text() === "With password") {
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: Checking the text value is not suitable. We should give flexibility to change them with i18n as well. Use some other method to check the condition.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    $('#domainSelector').change(function () {
+        var domain = $(this).val();
+        $("#domain").val(domain);
+    });
+    $('#verificationSelector').change(displayVals);
+});
+$(window).load(function () {
+    jQuery.validator.addMethod("noSpace", function (value, element) {
+        return value.indexOf(" ") < 0 ;
+    }, "Please don't enter spaces");
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: internationalization support missing.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            equalTo: "These passwords do not match.",
+            required: "Please re-enter the new password."
+        },
+        newPassword: {
+            required: "Required to provide a new password."
+        },
+        inputUsername:{
+            required: "Required to provide a username"
+        }
+
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: internationalization support missing in this block

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties [hunk 7]
```diff
@@ -33,6 +33,14 @@ users.add.select.with.email.or.phone=With email/phone verification
...
 users.add.password=Password
 users.confirm.password=Confirm Password
+user.add.error=Error while adding user.
+user.add.exist.error.message=Username is already taken. Please use a different name.
+user.add.failure=Failure!
+user.add.success.message=You have Succesfully added the user
+user.add.success=Success!
+user.add.error.retrieve.domain=Failed to retrieve the domain names.
+user.add.error.retrieve.primary.domain=Failed to retrieve the primary domain name.
+user.add.password.validate.error=Invalid Password.Please adhere to the password policy guidelines
```
Comment (nuwandiw) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties]: Add space after full-stop.

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/pom.xml [hunk 8]
```diff
@@ -107,6 +107,7 @@
                                     org.wso2.is.portal.user.client.api,
                                     org.wso2.is.portal.user.client.api.micro,
                                     org.wso2.is.portal.user.client.api.bean,
+                                    org.wso2.is.portal.user.client.api.util,
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/pom.xml]: add import version ranges

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        html: true,
+        content: $("#password_strength_wrap").html(),
+        placement: 'top',
+        trigger: 'focus keypress'
+    });
+    $("#newPassword").blur(function () {
+        $(".password_strength_meter .popover").popover("hide");
+    });
+}
+function showHidePassword(){
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js]: formatting issue

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            $(this).find('i').addClass("fw-hide");
+            $(this).parent().find('input[data-schemaformat=password]').attr('type', 'text');
+        }
+    });
+}
+function validateNewPassword() {
+    var password = $("#newPassword").val();
+    $.ajax({
+        type: "POST",
+        url: "/admin-portal/root/apis/passwordUtil-micro-service/validatePassword",
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js]: context /admin-portal/ should be configurable

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        data: {newPassword: password},
+        success: function (result) {
+            if (result === "true") {
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/fragments/password-generation/public/js/password-generation.js]: if result is  boolean check boolean comparison

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties [hunk 12]
```diff
@@ -33,6 +33,14 @@ users.add.select.with.email.or.phone=With email/phone verification
 profile=Profile
 users.add.password=Password
 users.confirm.password=Confirm Password
+user.add.error=Error while adding user.
+user.add.exist.error.message=Username is already taken. Please use a different name.
+user.add.failure=Failure!
+user.add.success.message=You have Succesfully added the user
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/lang/en_US.properties]: Succesfully --> successfully

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.hbs (hunk 1 above)
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.hbs]: i18n

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js [hunk 13]
```diff
@@ -2,17 +2,67 @@ function onGet(env) {
...
+    claimMap["http://wso2.org/claims/username"] = env.request.formParams['inputUsername'];
+    domain = env.request.formParams['domain'];
+    var registrationResult = userRegistration(claimMap, credentialMap, domain);
+    if (registrationResult.errorMessage) {
+        return {
+            domainNames: domainNames, primaryDomainName: primaryDomainName,
+            errorMessage: registrationResult.errorMessage
+        };
+    }
+    else {
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: formatting issue

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js [hunk 14]
```diff
@@ -2,17 +2,67 @@ function onGet(env) {
...
+    if (isRegexValidation === false) {
+        var passwordMinLength = PasswordPolicyConfigurationUtil.getPasswordMinLength();
+        var isNumbersIncluded = PasswordPolicyConfigurationUtil.isIncludeNumbers();
+        var isUpperCaseNeeded = PasswordPolicyConfigurationUtil.isIncludeUpperCase();
+        var isLowerCaseNeeded = PasswordPolicyConfigurationUtil.isIncludeLowerCase();
+        var isSpecialCharacterNeeded = PasswordPolicyConfigurationUtil.isIncludeSymbols();
+        sendToClient("result", {
+            passwordMinLength: passwordMinLength, isNumbersIncluded: isNumbersIncluded,
+            isUpperCaseNeeded: isUpperCaseNeeded, isLowerCaseNeeded: isLowerCaseNeeded,
+            isSpecialCharacterNeeded: isSpecialCharacterNeeded, regexvalidation: false
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: @nuwandiw regexvalidation should be in camel case.

Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: assign the result to a variable and send to client

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js [hunk 15]
```diff
@@ -25,8 +75,22 @@ function getPrimaryDomainName(env) {
...
+function userRegistration(claimMap, credentialMap, domain) {
+    try {
+        var userRegistrationResult = callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
+            "addUser", [claimMap, credentialMap, domain]);
+        return {userRegistration: userRegistrationResult, message: 'user.add.success.message'};
+    } catch (e) {
+        var message = "Error occurred while adding the user.";
+        Log.error(message, e);
+        return {
+            errorMessage: message
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: return i18n key

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js [hunk 16]
```diff
@@ -25,8 +75,22 @@ function getPrimaryDomainName(env) {
...
 }
+
+function userRegistration(claimMap, credentialMap, domain) {
+    try {
+        var userRegistrationResult = callOSGiService("org.wso2.is.portal.user.client.api.IdentityStoreClientService",
+            "addUser", [claimMap, credentialMap, domain]);
+        return {userRegistration: userRegistrationResult, message: 'user.add.success.message'};
+    } catch (e) {
+        var message = "Error occurred while adding the user.";
+        Log.error(message, e);
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/pages/users/add.js]: @nuwandiw remove unnecessary error logs

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js (hunk 5 above)
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: We need to discuss about supporting i18n in client side javascript

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js (hunk 4 above)
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: We need to discuss how to do fragment with UUF team

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+            onSuccess: function () {
+            },
+            onFailure: function (e) {
+            }
+        };
+        UUFClient.renderFragment("email-verification", fillingObject,
+            "accountVerificationMethod-area", "OVERWRITE", callbacks);
+    }
+}
+document.getElementById("inputUsername").onblur = function() {userNameExists()};
```
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: use jquery selector

File: portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js (hunk 6 above)
Comment (GayanM) [inline portal/apps/org.wso2.carbon.iam.admin-portal.app/src/main/public/js/add.js]: We need to discuss about supporting i18n in client side javascript

File: portal/client-services/org.wso2.is.portal.user.client.api/pom.xml [hunk 18]
```diff
@@ -86,9 +94,10 @@
             org.wso2.carbon.identity.claim.mapping.*;version="${org.wso2.carbon.identity.mgt.version.range}",
             org.wso2.carbon.uuf.spi.auth.*,
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/pom.xml]: add import version ranges

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+
+/**
+ * This is the micorservice to handle password utility services
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java]: Put . end of any comment

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+ */
+public class PasswordUtilityMicroService implements Microservice {
+
+    @POST
+    @Path("/validatePassword")
+    public Response isValidPassword(@Context Request password) throws UnsupportedEncodingException {
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java]: variable name 'password' is not correct

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        PasswordValidationService passwordValidationService = null;
+        boolean isValidPassword = false;
+        ByteBuffer fullContent = BufferUtil.merge(password.getFullMessageBody());
+        //TODO: Replace toString method, once a way to retrieve char[] without encoded values for special characters is
+        //found.
+        String passwordAsString = StandardCharsets.UTF_8.decode(fullContent).toString();
+        String decodedPassword = URLDecoder.decode(passwordAsString, StandardCharsets.UTF_8.toString());
+        char[] passwordCharArray = decodedPassword.toCharArray();
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java]: @ruwanta Please escalate this (password as char array)

File: portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        int index = IntStream.range(0, passwordCharArray.length).filter(i -> passwordCharArray[i] == '=')
+                .findFirst().orElse(-1);
+        char[] newArray = Arrays.copyOfRange(passwordCharArray, index + 1, passwordCharArray.length);
+        BundleContext bundleContext = FrameworkUtil.getBundle(PasswordValidationService.class).getBundleContext();
```
Comment (GayanM) [inline portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/micro/PasswordUtilityMicroService.java]: @ruwanta Please escalate this

File: tests/osgi-tests/src/test/java/org/wso2/is/portal/user/client/api/unit/test/IdentityStoreClientServiceTest.java [hunk 23]
```diff
@@ -122,6 +122,9 @@ public void testAddUserToDomain() throws UserPortalUIException {
         Assert.assertNotNull(user.getUserId(), "Invalid user unique id.");
 
         users.add(user);
+
+        boolean isUserExists = identityStoreClientService.isUserExist(userClaims, PRIMARY_DOMAIN);
+        Assert.assertTrue(isUserExists, "User does not exist in the given domain");
```
Comment (GayanM) [inline tests/osgi-tests/src/test/java/org/wso2/is/portal/user/client/api/unit/test/IdentityStoreClientServiceTest.java]: Message should reflect the idea when assert true

**Pull request merged**